### DISCUSSION TITLE
Best practice for having python and rust src?

### DISCUSSION THREAD

**Discussion by JackAshwell11** [PyO3/maturin#1376] [Q&A]
So I currently have a Python project which has a few slow components which I am converting to a Rust extensions. I want this Rust extension to be in the same directory as the Python source code and is just used to speed up the Python code. Here's the file structure:
```
root-dir/
├─ python_src/
│  ├─ __init__.py
│  ├─ python code here
├─ rust_src/
│  ├─ lib.rs
│  ├─ other rust code here
├─ pyproject.toml
├─ Cargo.toml
```
How can I attach documentation and a stub file to the Rust extension so that it can be accessed by the Python code in `python_src` (probably through installing the extension in the virtual environment)?

**Comment by messense** [comment]
> attach documentation

Doc-strings are turned into documentation automatically by pyo3.



> a stub file

see https://pyo3.rs/v0.17.3/python_typing_hints.html#including-pyi-files-in-your-pyo3maturin-build-package
